robottelo/orm.py

```diff
@@ -354,32 +354,31 @@ def read(self, auth=None, entity=None, attrs=None, ignore=()):
             for local_name, remote_name in entity.Meta.api_names.items():
                 attrs[local_name] = attrs.pop(remote_name)
 
-        # We must populate `entity`'s attributes from `attrs`.
+        # The server returns OneToOneFields as a hash of attributes, like so:
         #
-        # * OneToOneField names end with "_id"
-        # * OneToManyField names end with "_ids"
-        # * Other field names do not have any special name suffix.
+        #     {'name': 'Alice Hayworth', 'login': 'ahayworth', 'id': 1}
         #
-        # Well, that's the ideal. Unfortunately, the server often serves up
-        # weirdly structured or incomplete data. (See BZ #1122267)
+        # OneToManyFields are a list of hashes.
         for field_name, field_type in entity.get_fields().items():
             if field_name in ignore:
                 continue
             if isinstance(field_type, OneToOneField):
-                # `OneToOneField.entity` may be either a class or a string. For
-                # examples of this, look at a couple class definitions in
-                # module `robottelo.entities`. `_get_class` returns a class.
-                other_cls = _get_class(field_type.entity, 'robottelo.entities')
-                entity_id = attrs[field_name + '_id']
-                setattr(entity, field_name, other_cls(id=entity_id))
+                if attrs[field_name] is None:
+                    setattr(entity, field_name, None)
+                else:
+                    referenced_entity = _get_class(
+                        field_type.entity,
+                        'robottelo.entities'
+                    )(id=attrs[field_name]['id'])
+                    setattr(entity, field_name, referenced_entity)
             elif isinstance(field_type, OneToManyField):
                 other_cls = _get_class(field_type.entity, 'robottelo.entities')
-                entity_ids = attrs[field_name + '_ids']
-                setattr(
-                    entity,
-                    field_name,
-                    [other_cls(id=entity_id) for entity_id in entity_ids]
-                )
+                referenced_entities = [
+                    other_cls(id=referenced_entity['id'])
+                    for referenced_entity
+                    in attrs[field_name + 's']  # e.g. "users"
+                ]
+                setattr(entity, field_name, referenced_entities)
```

Comment:
The changes to this method are at the heart of what's going on in this PR. I've changed the format of data that `EntityReadMixin.read` expects to see by default. Formerly, the method expected to see data in this format:

    {
        'name': 'Alice',
        'supervisor_id': 3,
        'subordinate_ids': [4, 5],
    }

Now, the method expects to see data in this format:

    {
        'name': 'Alice',
        'supervisor': {'name': 'Bob', 'id': 3},
        'subordinates': [{'id': 4, 'name': 'Charlie'}, {'id': 5, 'name': 'Denise'}],
    }

The latter form is the most common form for receiving data from the API. The many `read` overrides in module `robottelo.entities` exist because there are still many exceptions to this rule.